events/controller/controller.go

```diff
@@ -0,0 +1,483 @@
+		logrus.Fields{
+			"ID":            n.controllerID,
+			"replica-count": n.kvReplicas,
+			"concurrency":   n.concurrency,
+		},
+	).Info("connected to event stream as controller")
+
+	return nil
+}
+
+// ConditionHandler is passed in by the caller to be invoked when the expected event is received
+type ConditionHandler interface {
+	Handle(ctx context.Context, condition *condition.Condition, publisher ConditionStatusPublisher) error
+}
+
+type ConditionHandlerFactory func() ConditionHandler
+
+// Handle events accepts a callback function to run when an event is fetched from the NATS JS.
+//
+// - The caller is expected to enclose any work and error handling for the work within the callback function.
+// - When the callback function returns no error, the event is marked as completed.
+// - When the callback function returns an ErrRetryHandler error, the corresponding event is placed back on the queue to be retried.
+func (n *NatsController) ListenEvents(ctx context.Context, chf ConditionHandlerFactory) error {
+	ctx, span := otel.Tracer(pkgName).Start(
+		ctx,
+		"ListenEvents",
+	)
+	defer span.End()
+
+	errListenEvents := errors.New("listen events error")
+	if chf == nil {
+		return errors.Wrap(errListenEvents, "expected valid ConditionHandlerFactor, got nil")
+	}
+
+	if n.stream == nil {
+		return errors.Wrap(errListenEvents, "connection not initialized, invoke Connect() before calling this method")
+	}
+
+	n.conditionHandlerFactory = chf
+
+	pullTicker := time.NewTicker(n.pullEventInterval)
+	defer pullTicker.Stop()
+
+Loop:
+	for {
+		select {
+		case <-pullTicker.C:
+			if n.concurrencyLimit() {
+				continue
+			}
+
+			if err := n.processEvents(ctx); err != nil {
+				return errors.Wrap(errListenEvents, err.Error())
+			}
+
+		case <-ctx.Done():
+			if n.dispatched > 0 {
+				continue
+			}
+
+			break Loop
+		}
+	}
+
+	return nil
+}
+
+// process event into a condition
+func (n *NatsController) processEvents(ctx context.Context) error {
+	pullCtx, cancel := context.WithTimeout(ctx, n.pullEventTimeout)
+	defer cancel()
+
+	errProcessEvent := errors.New("process event error")
+	msgs, err := n.stream.PullMsg(pullCtx, 1)
+	switch {
+	case err == nil:
+	case errors.Is(err, nats.ErrTimeout):
+		n.logger.WithFields(
+			logrus.Fields{"info": err.Error()},
+		).Trace("no new events")
+		return nil
+
+	default:
+		n.logger.WithFields(
+			logrus.Fields{"err": err.Error()},
+		).Warn("retrieving new messages")
+
+		metricsNATSError("pull-msg")
+		return errors.Wrap(errProcessEvent, err.Error())
+	}
+
+	for _, msg := range msgs {
+		if n.concurrencyLimit() {
+			return nil
+		}
+
+		// event status setter to keep the JS updated on our progress
+		eventAcknowleger := n.newNatsEventStatusAcknowleger(msg)
+
+		if ctx.Err() != nil {
+			eventAcknowleger.nak()
+
+			return errors.Wrap(errProcessEvent, ctx.Err().Error())
+		}
+
+		// setup the handle to query condition status
+		conditionStatusQueryor, err := n.NewNatsConditionStatusQueryor()
```

Comment:
The thought process there was to provide `processCondition` everything required to do its work since the method by itself is fairly large with the switch cases - we could split the switch cases into its own method like you suggest in a separate PR.